vscode/vscode/vscode.ml

```diff
@@ -1182,6 +1182,99 @@ module ProgressOptions = struct
       [@@js.builder]]
 end
 
+module DiagnosticSeverity = struct
+  type t =
+    | Error [@js 0]
+    | Hint [@js 1]
+    | Information [@js 2]
+    | Warning [@js 3]
+  [@@js.enum] [@@js]
+end
+
+module DiagnosticRelatedInformation = struct
+  include Class.Make ()
+
+  include
+    [%js:
+    val location : t -> Location.t [@@js.get]
+
+    val message : t -> string [@@js.get]
+
+    val make : location:Location.t -> message:string -> t
+      [@@js.new "vscode.DiagnosticRelatedInformation"]]
+end
+
+module DiagnosticTag = struct
+  type t =
+    | Unnecessary [@js 1]
+    | Deprecated [@js 2]
+  [@@js.enum] [@@js]
+end
+
+module Diagnostic = struct
+  include Class.Make ()
+
+  type string_or_int =
+    ([ `String of string
+     | `Int of int
+     ]
+    [@js.union])
+  [@@js]
+
+  let string_or_int_of_js js_val =
+    match Ojs.type_of js_val with
+    | "string" -> `String ([%js.to: string] js_val)
+    | "number" -> `Int ([%js.to: int] js_val)
+    | _ -> assert false
+
+  type code_target =
+    { value : string_or_int
+    ; target : Uri.t
+    }
+  [@@js]
+
+  type code =
+    ([ `String of string
+     | `Int of int
+     | `Targeted of code_target
+     ]
+    [@js.union])
+  [@@js]
+
+  let code_of_js js_val =
+    match Ojs.type_of js_val with
+    | "object" -> `Targeted ([%js.to: code_target] js_val)
+    | _ -> string_or_int_of_js js_val
```

Comment:
doesn't it make sense to inline `string_or_int_of_js` unless we want to put it in a larger scope to be reusable?

Reply:
`string_or_int_of_js` is being used (by gen_js_api) for the `value` field of `code_target`.
So the function is already being reused locally.

Reply:
We could add these kind of types to the `interop` module, but I don't see the need for it yet.